feat(format): output references

dbanksdesign opened the pull request:
*Issue #, if available:* #17 #396

*Description of changes:* Allow for variable references in output files. These changes will make it easier to use the *name* of a referenced token rather than the resolved value, so the output will keep the aliases/references intact.

* Adding `resolveReference()` and `usesReference()` methods on the dictionary object that is passed to formats and actions. This will allow custom formats and actions to test and resolve references the same way style dictionary does internally, allowing you to get the transformed name of a token for instance.
* Adding a `outputReferences` configuration on SCSS, CSS, and Less variables, as well as `ios/class.swift`, `flutter/class.dart`, and `android/resources` formats to make use of this
* Adding a 'output-references' example to show how to use these new methods
* Added documentation about this feature
* Added integration testing to properly test this


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

chazzmoney commented:
WOOHOO! #17

chazzmoney commented:
I really like that this empowers people on the route of #268, which is something that I have wanted for SD for a LONG time (even well before I created that issue).

I think it is super valuable that people can customize SD for how **they** think about tokens. Even so, it encourages base definitions in one place and output definitions in another.

It also enables client side debugging/tooling to have strong access to the underlying reference structure that powers SD.

Really really excited about this one.

chazzmoney commented:
Two improvements to consider:

1. I think this deserves documentation catering to the absolute newest / least experienced SD user. This is a very important / useful thing and making sure people know it is available and can understand how to use it easily will help them use SD to its fullest.
2. Based on the usefulness to downstream tooling, we may want to make this on by default... this will be included in a major version/breaking change so it might be ok... we should probably @ some of our most active folks here to get their opinion.
3. The configuration options here should likely be able to be set at three levels: global, included file, and property (maybe even inheritably so)

chazzmoney commented:
Also... we should probably update all default formats to use variables where appropriate.

It will be a lot of work, but it is weird to include this without updating all of them.

didoo commented on lib/common/formats.js:
```diff
@@ -55,9 +55,20 @@ function formattedVariables(format, properties) {
+      if (useVariables && dictionary.usesReference(prop.original.value)) {
+        var ref = dictionary.getReference(prop.original.value).name;
+        if (format === 'css') {
+          value = `var(${prefix}${ref})`
```
nitpicking here, but in the output it generates something like:
```
:root {
  --token-color-error: var(  --token-color-generic-red);
}
```
and that extra space inside the var() is... annoying :)

I think more correctly the `prefix` value at line 46 should be `--` and not `  --` (formally, more correct) and then we should add two extra spaces in output when we generate the line (`var to_ret_prop = prefix + prop.name + ': ';`)

didoo commented on lib/common/formats.js:
```diff
@@ -55,9 +55,20 @@ function formattedVariables(format, properties) {
+        if (format === 'css') {
+          value = `var(${prefix}${ref})`
+        } else {
+          value = prefix + ref;
```
for consistency with two lines above I would do:
```suggestion
          value = `${prefix}${ref}`;
```

didoo commented on examples/advanced/variables-in-outputs/README.md:
````diff
@@ -0,0 +1,48 @@
+
+The `sd.config.js` file has everything you need to see. The tokens included in this example are just enough to show these new techniques.
+
+```javascript
````
```suggestion
Here is an example that shows how to get an alias's name within a custom format:
```javascript
```

didoo commented on lib/utils/references/resolveReference.js:
```diff
@@ -0,0 +1,36 @@
+    return;
+  }
+
+  for (i = 0; i < path.length; i++) {
```
now that it's been decided to move away from lodash, and the "utils" folder has started to contain lodash-like functions, do you think we should isolate this as `deepFind` function there?

didoo commented:
First of all, amazing PR @dbanksdesign, really great idea 🙌.

Now, I did a quick code review and added a few comments there. I have some other thoughts here:
- I am not entirely sure about the name `useVariables`. maybe there is something that could be more explicit on what it does? I was thinking something like `preserve[References/Variables]` or `use[Variables|References]InOutput`
- agree with @chazzmoney that we should consider also other less-obvious default formats that may support this kind of referencing in output (which one? JS clearly yes, in some form; but what about others like formats for iOS or Android?)
- regarding setting it on by default for version 3.0, agree as well, makes sense (unless there are downsides we're overlooking)
- regarding the three levels configuration, not sure about it. why do you think we need this extra layer of complexity in the codebase? if you can achieve the result adding a set of `useVariables: true` wherever you need it for the output, I don't think is much of a burden for the final user, and in any case, you probably do it once at the initial setup and then forget.
- regarding the documentation, there are other parts that we may want to highlight for the new user (eg. my colleague didn't realize the input format for the tokens could be directly JavaScript code, not only JSON) so maybe this could be done in a separate task?

Let me know if I can help somehow on this.

dbanksdesign commented:
@didoo I would love some help with this! I think we need some more tests, more documentation, and updating applicable formats to use this new code. Whichever you'd like to help out on would be greatly appreciated. I also have some time tomorrow I can spend on this and you can take a look.

dbanksdesign commented:
Big update:
* Changed the option name from 'useVariables' to 'keepReferences' so it better matches the methods
* Moved the option to the options object in a file, to be consistent with other file-level configuration like `showFileHeader`
* Updated `android/resources` format to handle `keepReferences`
* Updated `scss/variables` format to handle `keepReferences`. In order to do this, because Sass is imperative, we are sorting the `allProperties` array so all references come last.
* Started to add integration testing to better test the output of formats. 
* Cleaned up some things in our testing

dbanksdesign commented:
Another update:
* Changed option name to `outputReferences` and moved it to the `options` object on a file
* Added Swift and Flutter support and integration tests
* Started some documentation

chazzmoney commented on docs/formats.md:
````diff
@@ -73,6 +73,54 @@ A special file configuration is `filter`, which will filter the tokens before th
+  --color-danger: var(--color-red);
+  --color-error: var(--color-danger);
+}
+```
````
I love this as an example. You might contrast it with what happens if outputReferences is false.

chazzmoney commented:
Nice work. Seriously. LGTM! :shipit:

dbanksdesign commented:
Update:
* Added documentation in the config section about this
* Platform level configuration can now define an `options` object with the same options as a file object, and the options cascade
* Renamed the example

uptonking commented:
@dbanksdesign Is there any way to support generate css variables with fallback values?
tokens source: 
```json
{ 
  "color": {
    "background": {
      "primary": { 
		"value": "{color.palette.neutral.0.value}" ,
		"fallback":"red"
		}
    }
  }
}
```

tokens output: 
```css
/** simple form */
--color-background-primary: var(--my-var, red);

/* complex var may not be necessary */
--color-background-primary: var(--my-var, var(--my-background, pink));
```

https://github.com/amzn/style-dictionary/issues/544 this issue seems similar.
